src/verify.rs

```diff
@@ -0,0 +1,615 @@
+use bls12_381::hash_to_curve::{ExpandMsgXmd, HashToCurve};
+use bls12_381::{
+    multi_miller_loop, G1Affine, G1Projective, G2Affine, G2Prepared, G2Projective, Gt,
+};
+use serde::{Deserialize, Deserializer};
+use sha2::{Digest, Sha256};
+use std::ops::Neg;
+use thiserror::Error;
+
+#[derive(Deserialize, Debug, PartialEq, Clone)]
+pub struct Beacon {
+    #[serde(alias = "round")]
+    pub round_number: u64,
+    #[serde(with = "hex")]
+    pub randomness: Vec<u8>,
+    #[serde(with = "hex")]
+    pub signature: Vec<u8>,
+    #[serde(default, with = "hex")]
+    pub previous_signature: Vec<u8>,
+}
+
+const DST_G1: &str = "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_";
+const DST_G2: &str = "BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_";
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum SchemeID {
+    PedersenBlsChained,
+    PedersenBlsUnchained,
+    UnchainedOnG1,
+    UnchainedOnG1RFC9380,
+}
+
+impl<'de> Deserialize<'de> for SchemeID {
+    fn deserialize<D>(deserializer: D) -> Result<SchemeID, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = Deserialize::deserialize(deserializer)?;
+        match s {
+            "pedersen-bls-chained" => Ok(SchemeID::PedersenBlsChained),
+            "pedersen-bls-unchained" => Ok(SchemeID::PedersenBlsUnchained),
+            "bls-unchained-on-g1" => Ok(SchemeID::UnchainedOnG1),
+            "bls-unchained-g1-rfc9380" => Ok(SchemeID::UnchainedOnG1RFC9380),
+            _ => Err(serde::de::Error::unknown_variant(
+                s,
+                &[
+                    "pedersen-bls-chained",
+                    "pedersen-bls-unchained",
+                    "bls-unchained-on-g1",
+                    "bls-unchained-g1-rfc9380",
+                ],
+            )),
+        }
+    }
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum VerificationError {
+    #[error("unchained schemes cannot contain a `previous_signature`")]
+    UnchainedHasPreviousSignature,
+    #[error("chained beacons must have a `previous_signature`")]
+    ChainedBeaconNeedsPreviousSignature,
+    #[error("invalid signature length")]
+    InvalidSignatureLength,
+    #[error("invalid public key")]
+    InvalidPublicKey,
+    #[error("message can't be empty")]
+    EmptyMessage,
+    #[error("signature verification failed")]
+    SignatureFailedVerification,
+    #[error("the randomness for the beacon did not match the signature")]
+    InvalidRandomness,
+}
+
+pub fn verify_beacon(
+    scheme_id: &SchemeID,
+    public_key: &[u8],
+    beacon: &Beacon,
+) -> Result<(), VerificationError> {
+    if Sha256::digest(&beacon.signature).to_vec() != beacon.randomness {
+        return Err(VerificationError::InvalidRandomness);
+    }
```

Comment:
This would fail for the `PedersenBlsChained` mode, no? Add some test vectors :P

Reply:
why would it fail for chained? ;o this is just checking the `randomness` == `hash(signature` which is the same for chained and unchained beacons afaiu. Also there is a test vector for default beacons

Reply:
Ah right, I was thinking of the message we're signing